File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));


        while (true) {
            String[] s = br.readLine().split("");
            if (s[0].equals(".")) break;
            String ans = "yes";
            Stack<String> stack = new Stack<>();

            for(String x : s) {
                if (x.equals("(") || x.equals("[")) {
                    stack.push(x);
                } else if (x.equals(")")){
                    if (!stack.isEmpty() && stack.peek().equals("(")){
                        stack.pop();
                    } else {
                        ans = "no";
                        break;
                    }
                } else if (x.equals("]")) {
                    if (!stack.isEmpty() && stack.peek().equals("[")){
                        stack.pop();
                    } else {
                        ans = "no";
                        break;
                    }
                } else if (x.equals(".")) {
                    break;
                }
            }

            if (!stack.isEmpty()) ans = "no";
            bw.write(ans + "\n");
        }
        bw.flush();
        bw.close();
        br.close();
    }
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.kt
import java.util.*
import java.io.*
import java.util.Stack

/*
백준 4949번
 */

fun main(args: Array<String>) {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    loop@ while(true) {
        val line = br.readLine()
        if (line == ".") break
        val token = StringTokenizer(line, ".")
        for (t in token) {
            val str = t.toString()
            val stack = Stack<Char>()
            if (str.contains("(") || str.contains("[")
                || str.contains(")") || str.contains("]")
            ) {
                for (s in str) {
                    when (s) {
                        '(', '[' -> stack.push(s)
                        ')' -> {
                            if (!stack.isEmpty() && stack.peek() == '(') stack.pop()
                            else {
                                bw.write("no\n")
                                continue@loop
                            }
                        }
                        ']' -> {
                            if (!stack.isEmpty() && stack.peek() == '[') stack.pop()
                            else {
                                bw.write("no\n")
                                continue@loop
                            }
                        }
                    }
                }
            }
            bw.write(if (stack.isEmpty()) "yes\n" else "no\n")
        }
    }
    bw.flush()
    bw.close()
}